storage/replica.go

```diff
@@ -341,11 +341,27 @@ func (r *Replica) requestLeaderLease(timestamp roachpb.Timestamp) error {
 	// Note that the command itself isn't traced, but usually the caller
 	// waiting for the result has an active Trace.
 	errChan, pendingCmd := r.proposeRaftCommand(r.context(), ba)
-	if err := <-errChan; err != nil {
-		return err
+
+	// It does no good to block here for longer than the proposed lease's
+	// lifetime. If the command takes that long to commit we are probably
+	// partitioned away and we can conclude that another node has taken
+	// the lease.
+	deadline := time.After(duration)
```

Comment:
I agree that we shouldn't be using the client's context here, but using a freshly deadlined `context.Context` is strictly better in that it would be passed into `proposeRaftCommand` above, which allows all moving parts up the callstack to cancel the request early.